video.c

```diff
@@ -747,73 +873,130 @@ render_line(uint16_t y)
 	render_layer_line(0, eff_y);
 	render_layer_line(1, eff_y);
 
-	for (uint16_t x = 0; x < SCREEN_WIDTH; x++) {
-		uint8_t r;
-		uint8_t g;
-		uint8_t b;
-		if (out_mode == 0) {
-			// video generation off
-			// -> show blue screen
-			r = 0;
-			g = 0;
-			b = 255;
-		} else {
-			uint8_t col_index = 0;
-			if (x < hstart || x > hstop || y < vstart || y > vstop) {
-				col_index = border_color;
-			} else {
-				int eff_x = 1.0 / hscale * (x - hstart);
-				uint8_t spr_col_index = sprite_line_empty ? 0 : sprite_line_col[eff_x];
-				uint8_t spr_zindex = sprite_line_z[eff_x];
-				uint8_t l1_col_index = layer_line_empty[0] ? 0 : layer_line[0][eff_x];
-				uint8_t l2_col_index = layer_line_empty[1] ? 0 : layer_line[1][eff_x];
-				switch (spr_zindex) {
+	uint8_t col_line[SCREEN_WIDTH];
+
+	if (video_palette.dirty) {
+		refresh_palette();
+	}
+
+	// If video output is enabled, calculate color indices for line.
+	if (out_mode != 0) {
+		uint8_t spr_col_index[LAYER_PIXELS_PER_ITERATION];
+		uint8_t l1_col_index[LAYER_PIXELS_PER_ITERATION];
+		uint8_t l2_col_index[LAYER_PIXELS_PER_ITERATION];
+		uint8_t spr_zindex[LAYER_PIXELS_PER_ITERATION];
+
+		memset(spr_col_index, 0, sizeof(spr_col_index));
+		memset(l1_col_index, 0, sizeof(l1_col_index));
+		memset(l2_col_index, 0, sizeof(l2_col_index));
+		memset(spr_zindex, 0, sizeof(spr_zindex));
+
+		// Calculate color without border.
+		for (uint16_t x = 0; x < SCREEN_WIDTH; x+=LAYER_PIXELS_PER_ITERATION) {
+			uint8_t col_index[LAYER_PIXELS_PER_ITERATION];
+			memset(col_index, 0, sizeof(col_index));
+
+			int eff_x[LAYER_PIXELS_PER_ITERATION];
+			for (int i = 0; i < LAYER_PIXELS_PER_ITERATION; ++i) {
+				eff_x[i] = 1.0 / hscale * (x + i - hstart);
```

Comment:
`eff_x[i] = (reg_composer[1] * (x + i - hstart)) >> 7;` would be more efficient, since it avoids double division

Reply:
same trick goes for `eff_y`